File: examples/ts/ec-unit/ECUnit.runar.ts
import {
  SmartContract, assert, len,
  ecAdd, ecMul, ecMulGen, ecNegate, ecOnCurve,
  ecEncodeCompressed, ecMakePoint, ecPointX, ecPointY,
} from 'runar-lang';
import type { ByteString } from 'runar-lang';

class ECUnit extends SmartContract {
    readonly pubKey: ByteString;
    constructor(pubKey: ByteString) {
        super(pubKey);
        this.pubKey = pubKey;
    }

    public testOps() {
        const g = ecMulGen(1n);
        assert(ecOnCurve(g));
        const neg = ecNegate(g);
        assert(ecOnCurve(neg));
        const doubled = ecMul(g, 2n);
        assert(ecOnCurve(doubled));
        const sum = ecAdd(g, g);
        assert(ecOnCurve(sum));
        // P + (-P) is the point at infinity, which affine x||y cannot encode:
        // ecAdd returns the all-zero blob from a SUCCESSFUL script (as does
        // ecMul for any k = 0 mod n). ecOnCurve is the only way to notice,
        // so it must reject it. Selecting the tangent on px == qx alone
        // returned 2G here instead: on-curve, plausible, and accepted.
        const infPt = ecAdd(g, neg);
        assert(!ecOnCurve(infPt));
        const x = ecPointX(g);
        const y = ecPointY(g);
        const rebuilt = ecMakePoint(x, y);
        assert(ecOnCurve(rebuilt));
        const compressed = ecEncodeCompressed(g);
        assert(len(compressed) == 33n);
        assert(true);
    }
}

File: examples/go/ec-unit/ECUnit.runar.go
package contract

import runar "github.com/icellan/runar/packages/runar-go"

// ECUnit -- Unit-style exercises for the secp256k1 EC built-ins.
type ECUnit struct {
	runar.SmartContract
	PubKey runar.ByteString `runar:"readonly"`
}

// TestOps exercises ecMulGen, ecOnCurve, ecNegate, ecMul, ecAdd, ecPointX,
// ecPointY, ecMakePoint, and ecEncodeCompressed.
func (c *ECUnit) TestOps() {
	g := runar.EcMulGen(1)
	runar.Assert(runar.EcOnCurve(g))
	neg := runar.EcNegate(g)
	runar.Assert(runar.EcOnCurve(neg))
	doubled := runar.EcMul(g, 2)
	runar.Assert(runar.EcOnCurve(doubled))
	sum := runar.EcAdd(g, g)
	runar.Assert(runar.EcOnCurve(sum))
	// P + (-P) is the point at infinity: EcAdd returns the all-zero blob from
	// a SUCCESSFUL script, and EcOnCurve must reject it.
	infPt := runar.EcAdd(g, neg)
	runar.Assert(!runar.EcOnCurve(infPt))
	x := runar.EcPointX(g)
	y := runar.EcPointY(g)
	rebuilt := runar.EcMakePoint(x, y)
	runar.Assert(runar.EcOnCurve(rebuilt))
	compressed := runar.EcEncodeCompressed(g)
	runar.Assert(runar.Len(compressed) == 33)
	runar.Assert(true)
}
